cmake/DefaultTargetProperties.cmake

```diff
@@ -0,0 +1,129 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2013-2022 The Foundry Visionmongers Ltd
+
+function(set_default_target_properties target_name)
+    #-------------------------------------------------------------------
+    # C++ standard
+
+    # Minimum C++ standard
+    target_compile_features(${target_name} PRIVATE cxx_std_14)
+
+    set_target_properties(
+        ${target_name}
+        PROPERTIES
+        # Ensure the proposed compiler supports our minimum C++
+        # standard.
+        CXX_STANDARD_REQUIRED ON
+        # Disable compiler extensions. E.g. use -std=c++11  instead of
+        # -std=gnu++11.  Helps limit cross-platform issues.
+        CXX_EXTENSIONS OFF
+    )
+
+    #-------------------------------------------------------------------
+    # Compiler warnings
+
+    include(CompilerWarnings)
+    set_default_compiler_warnings(${target_name})
+
+
+    #-------------------------------------------------------------------
+    # Potentially enable sanitizers.
+    include(Sanitizers)
+    enable_sanitizers(project_options)
+
+
+    #-------------------------------------------------------------------
+    # Symbol visibility
+
+    # Hide symbols from this library by default.
+    set_target_properties(
+        ${target_name}
+        PROPERTIES
+        C_VISIBILITY_PRESET hidden
+        CXX_VISIBILITY_PRESET hidden
+        VISIBILITY_INLINES_HIDDEN YES
+    )
+
+    # Hide all symbols from external statically linked libraries.
+    if (CMAKE_CXX_COMPILER_ID MATCHES ".*Clang" OR CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+        target_link_options(${target_name} PRIVATE -Wl,--exclude-libs,ALL)
+    endif ()
+
+    # Use C++11 ABI by default, but allow disabling.
+    if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND
+        CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 5.0)
+        option(GLIBCXX_USE_CXX11_ABI "For gcc, use the new C++11 library ABI" ON)
+        if (GLIBCXX_USE_CXX11_ABI)
+            target_compile_definitions(${target_name} PRIVATE _GLIBCXX_USE_CXX11_ABI=1)
+        else ()
+            target_compile_definitions(${target_name} PRIVATE _GLIBCXX_USE_CXX11_ABI=0)
+        endif ()
+    endif ()
+
+    # TODO(DF): POSITION_INDEPENDENT_CODE required for Python module? See OCIO.
+
+    #-------------------------------------------------------------------
+    # RPATH wrangling.
+
+    # With the 'usual' install path structure in mind, search from the
+    # bin directory (i.e. a binary loading a dynamic library) and then
+    # from the current directory (i.e. dynamic library loading another
+    # dynamic library).
+    if (APPLE)
+        set(rpath "@loader_path/../${CMAKE_INSTALL_LIBDIR};@loader_path")
+    else ()
+        set(rpath "$ORIGIN/../${CMAKE_INSTALL_LIBDIR};$ORIGIN")
```

Comment:
I'd be surprised this wasn't the default.

Reply:
Just a heads up: Beware of runpath (generally the default on compilers in non-CentOS distros). It doesn't behave the same than with rpath. It depends how the compiler was configured as to what you get, and what you pass to the linker.

It'll only really bite you for shared libraries that load shared libraries. It may not find them, even if they're in the same directory. It may not bite you, but bite your clients instead.

`-Wl,--disable-new-dtags` is your friend here.

Couple of references:
https://tipsfordev.com/run-time-linker-issue-error-while-loading-shared-libraries
https://stackoverflow.com/questions/58997230/cmake-project-fails-to-find-shared-library/60126535#60126535

RUNPATH vs RPATH is on my research list, as we'll be hitting it when we change build distro in the future.

Reply:
Ah yes, RPATH vs RUNPATH is fun. 

The difference in transitive search path behaviour bit me in the previous incarnation of OpenAssetIO (FnAssetAPI) - its remarkable how poorly documented (or at least highlighted) this massive difference is!

The transitive search path feature of RPATH sounds like it would make life easier, however buried in the  [`ld` docs](https://man7.org/linux/man-pages/man8/ld.so.8.html):

> Use of DT_RPATH is deprecated

That, in addition to it taking precedence over a user-defined `LD_LIBRARY_PATH` (unlike `RUNPATH`), makes me unsure if its a good idea to use `RPATH`s (i.e. `--disable-new-dtags`), at least by default.

Perhaps in practice, regardless of notional deprecation, downstream adopters would prefer RPATH?  That's certainly something you can answer much better than me...?

Reply:
Also interesting that you imply CentOS defaults to `RPATH`?  If that's the case, then for cross-distro compatibility that would force our hand.

Reply:
Historically, RPATH was the thing. Then RUNPATH was introduced due to limitations/restrictions of RPATH.
In CentOS 7 (and below), their GCC's (including devtoolsets) are compiled without the new dtags flag, so default to using RPATH. On recent Ubuntu's their GCC defaults to dtags enabled, thus using RUNPATH. I believe the new gcctoolset's on RHEL also do that.

RPATH works very well I find. RUNPATH gets in the way most of the time. But my opinion at the moment.

Reply:
Forced RPATH (dtags) setting in c9dac6fc